plugin/output/s3/s3_internals.go

```diff
@@ -0,0 +1,175 @@
+		}
+	}
+
+	return defaultClient, minioClients, nil
+}
+
+func (p *Plugin) getDirs(outPlugCount int) map[string]string {
+	// dir for all bucket files.
+	dir, _ := filepath.Split(p.config.FileConfig.TargetFile)
+
+	targetDirs := make(map[string]string, outPlugCount)
+	targetDirs[p.config.DefaultBucket] = dir
+	// multi_buckets from config are sub dirs on in Config.FileConfig.TargetFile dir.
+	for _, singleBucket := range p.config.MultiBuckets {
+		targetDirs[singleBucket.Bucket] = filepath.Join(dir, staticBucketDir, singleBucket.Bucket) + dirSep
+	}
+	return targetDirs
+}
+
+func (p *Plugin) getFileNames(outPlugCount int) map[string]string {
+	fileNames := make(map[string]string, outPlugCount)
+
+	_, f := filepath.Split(p.config.FileConfig.TargetFile)
+	p.fileExtension = filepath.Ext(f)
+
+	mainFileName := f[0 : len(f)-len(p.fileExtension)]
+	fileNames[p.config.DefaultBucket] = mainFileName
+	for _, singleB := range p.config.MultiBuckets {
+		fileNames[singleB.Bucket] = singleB.Bucket
+	}
+	return fileNames
+}
+
+// Try to create buckets from dirs lying in dynamic_dirs route
+func (p *Plugin) createPlugsFromDynamicBucketArtifacts(targetDirs map[string]string) {
+	dynamicDirsPath := filepath.Join(targetDirs[p.config.DefaultBucket], dynamicBucketDir)
+	dynamicDir, err := ioutil.ReadDir(dynamicDirsPath)
+	if err != nil {
+		p.logger.Infof("%s doesn't exist, willn't restore dynamic s3 buckets", err.Error())
+		return
+	}
+	for _, f := range dynamicDir {
+		if !f.IsDir() {
+			continue
+		}
+		created := p.tryRunNewPlugin(f.Name())
+		if created {
+			p.logger.Infof("dynamic bucket '%s' retrieved from artifacts", f.Name())
+		} else {
+			p.logger.Infof("can't retrieve dynamic bucket from artifacts: %s", err.Error())
+		}
+	}
+}
+
+func (p *Plugin) createOutPlugin(bucketName string) (*file.Plugin, error) {
+	exists, err := p.clients[bucketName].BucketExists(bucketName)
+	if err != nil {
+		return nil, fmt.Errorf("could not check bucket %s, error: %w", bucketName, err)
+	}
+	if !exists {
+		return nil, &ErrCreateOutputPluginNoSuchBucket{bucketName: bucketName}
+	}
+
+	anyPlugin, _ := file.Factory()
+	outPlugin := anyPlugin.(*file.Plugin)
+	outPlugin.SealUpCallback = p.addFileJobWithBucket(bucketName)
+
+	return outPlugin, nil
+}
+
+func (p *Plugin) startPlugins(Params *pipeline.OutputPluginParams, outPlugCount int, targetDirs, fileNames map[string]string) error {
+	outPlugins := make(map[string]*file.Plugin, outPlugCount)
+	outPlugin, err := p.createOutPlugin(p.config.DefaultBucket)
+	if err != nil {
+		return err
+	}
+	outPlugins[p.config.DefaultBucket] = outPlugin
+	p.logger.Infof("bucket %s exists", p.config.DefaultBucket)
+
+	p.createPlugsFromDynamicBucketArtifacts(targetDirs)
+	// If multi_buckets described on file.d config, check each of them as well.
+	for _, singleBucket := range p.config.MultiBuckets {
+		outPlugin, err := p.createOutPlugin(singleBucket.Bucket)
+		if err != nil {
+			return err
```

Comment:
Do we need to stop other already created plugins in case of err?

Reply:
It'll go up to main func and die there. I think we are ok.